modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala

```diff
@@ -48,5 +58,58 @@ object ConstructionService {
 
       NonEmptyList.fromList(accountIdentifiers)
     }
-  }
+
+    def transactionToOperations(transaction: Transaction): Either[String, NonEmptyList[Operation]] = {
+      refineV[Not[Equal[0L]]](transaction.amount.value.value).flatMap { amount =>
+        val operations: Either[String, List[Operation]] = refineV[Not[Equal[0L]]](-1L * transaction.amount.value.value).map { negAmount =>
+            val negativeOperation = Operation(
+              OperationIdentifier(OperationIndex(0L)),
+              none,
+              OperationType.Transfer,
+              none,
+              AccountIdentifier(transaction.destination, none).some,
+              Amount(AmountValue(negAmount), DAG).some
+            )
+
+            val positiveOperation = Operation(
+              OperationIdentifier(OperationIndex(1L)),
+              none,
+              OperationType.Transfer,
+              none,
+              AccountIdentifier(transaction.destination, none).some,
+              Amount(AmountValue(amount), DAG).some
+            )
+
+            List(negativeOperation, positiveOperation)
+        }
+
+        operations.flatMap{ operationList =>
+          NonEmptyList.fromList(operationList) match {
+            case Some(value) => value.asRight
+            case None => "No valid operations found.".asLeft
+          }
+        }
+      }
+    }
+
+  def parseSignedTransaction(hex: Hex): F[Either[MalformedTransaction.type, ConstructionParse.Response]] = 
+    KryoSerializer[F].deserialize[Signed[Transaction]](hex.toBytes).traverse { signedTransaction =>
+    transactionToOperations(signedTransaction).traverse { operationsList =>
+      signedTransaction.proofs.toNonEmptyList.traverse(signature => signature.id.hex.toPublicKey).map { proofs =>
+        val proofAccountIdentifiers = proofs.map(_.toAddress).map { address =>
+          AccountIdentifier(address, none)
+        }
+        
+        ConstructionParse.Response(operationsList.toList, proofAccountIdentifiers.toList.some)
+      }
+    }.map(_.left.map(_ => MalformedTransaction))
+  }.map(_.left.map(_ => MalformedTransaction).flatten)
+
+  def parseUnsignedTransaction(hex: Hex): Either[MalformedTransaction.type, ConstructionParse.Response] = 
+    KryoSerializer[F].deserialize[Transaction](hex.toBytes).traverse { transaction =>
+    transactionToOperations(transaction).map { operationsList =>
+        ConstructionParse.Response(operationsList.toList, none)
+      }
+    }.map(_.left.map(_ => MalformedTransaction)).left.map(_ => MalformedTransaction).flatten
```

Comment:
```scala
      KryoSerializer[F]
        .deserialize[Transaction](hex.toBytes)
        .flatMap(transactionToOperations)
        .map(ops => ConstructionParse.Response(ops.toList, none))
        .leftMap(_ => MalformedTransaction)
```

Reply:
yep, but the service should not return an API response

Reply:
Changed the code as pointed out by @kpudlik . Regarding @marcinwadon 's comment, I can change the return object to a case class made for that purpose, but I still have no idea how to convert `F[Either[MalformedTransaction.type, ConstructionParse.Response]]` and `Either[MalformedTransaction.type, ConstructionParse.Response]` into `F[Response[F]]`.